### PR TITLE
Feat/rank histogram metric

### LINKED ISSUE #540: Add rank histogram metric for ensemble calibration assessment
## Summary

Add `rank_histogram()` function to `metrics.py` for ensemble forecast calibration diagnostics.

## Motivation

Rank histograms are standard diagnostic tools for probabilistic forecasts, referenced in issue #62. They show if an ensemble is well-calibrated (flat), underdispersive (U-shaped), or overdispersive (dome-shaped).

Complements sample-based CRPS being added in #226.

## Implementation

```python
def rank_histogram(ensemble_pred, target, mask=None, num_bins=None):
    """
    ensemble_pred: (M, ..., N, d_state)
    target: (..., N, d_state)
    Returns: (num_bins, d_state) histogram counts
    """
```

References
Issue #62 (Graph-EFM merge)
Hamill (2001): "Interpretation of Rank Histograms for Verifying Ensemble Forecasts"

### PR DESCRIPTION
**Pull request by aryannzzz**
## Describe your changes

Adds rank_histogram() function to [metrics.py](http://metrics.py/) for assessing ensemble forecast calibration. Computes which rank bin each observation falls into among sorted ensemble members. For M members there are M+1 bins.
Returns (num_bins, d_state) histogram counts per variable.
Intentionally not added to DEFINED_METRICS since it takes ensemble_pred with shape (M, ..., N, d_state) rather than pred, its a diagnostic tool, not a training loss.

## Issue Link

closes #540

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)

- [x] ✨ New feature (non-breaking change that adds functionality)

- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)

- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).

- [x] I have performed a self-review of my code

- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values

- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code

- [x] I have updated the [README]([README.MD](http://readme.md/)) to cover introduced code changes

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).

- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:

- [x] the code is readable

- [x] the code is well tested

- [x] the code is documented (including return types and parameters)

- [x] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section

  reflecting type of change (add section where missing):

  - *added*: when you have added new functionality

  - *changed*: when default behaviour of the code has been changed

  - *fixes*: when your contribution fixes a bug

  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch

- [ ] the tests pass

- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.

- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)

- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by sadamov** [comment]
I am personally a big fan of the rank/PIT histograms. But I am not sure if it should be a part of the metrics in neural-lam. As you mention it is more of a diagnostic tool. I have often seen it being used within separate verification software, often presented as barplots. The already existing SSR/CRPS do also measure dispersiveness to some degree. so unless there is a good reason we also need them here I would rather use existing verification software on neural-lam zarr output

File: neural_lam/metrics.py [hunk 1]
```diff
@@ -226,6 +226,63 @@ def crps_gauss(
...
+    # Move ensemble dim to last for easier comparison
+    # (M, ..., N, d_state) -> (..., N, d_state, M)
+    ensemble_transposed = ensemble_pred.movedim(0, -1)
+
+    # Expand target to match: (..., N, d_state) -> (..., N, d_state, 1)
+    target_expanded = target.unsqueeze(-1)
+
+    # Count how many ensemble members are below the target
+    # ranks are in {0, 1, ..., M}, representing M+1 possible positions
+    ranks = (ensemble_transposed < target_expanded).sum(dim=-1)  # (..., N, d_state)
```
Comment (kshirajahere) [inline neural_lam/metrics.py:263]: This rank calculation is not correct when the target ties with one or more ensemble members. Using a strict `<` count forces all ties into the lower admissible bin, which biases the histogram for quantized or repeated-valued forecasts/targets. Rank/PIT diagnostics usually need either randomized tie-breaking or at least an explicit documented tie convention, otherwise a calibrated ensemble can still produce a distorted histogram here.

File: tests/test_metrics.py
Comment (kshirajahere) [inline tests/test_metrics.py:1]: I think this test coverage is still missing the most important edge case for this implementation: ties between `target` and one or more ensemble members. Since the current rank logic is based on a strict `<` comparison, a tie-focused regression test would help pin down whether that behavior is intended or whether tie handling needs to be adjusted.

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Add `rank_histogram` metric for ensemble forecast calibration diagnostics [\#540](https://github.com/mllam/neural-lam/pull/540)
```
Comment (kshirajahere) [inline CHANGELOG.md:12]: Minor changelog nit: this entry says #540, but #540  is the issue, not a PR. Should have the PR #541 instead

**Review by kshirajahere — COMMENTED** [review]
Hey, was browsing the open Project 3 / probabilistic-eval work and this one seemed interesting, since rank/PIT diagnostics are a natural part of the ensemble-forecasting path.

I like the direction overall, but I think there are still a few things to tighten before this is review-ready.

**Comment by sadamov** [comment]
Closing this in favour of #649, which consolidates the probabilistic metrics + ensemble plumbing track for v0.8.0. The rank histogram (Talagrand) metric will be folded into the consolidated PR. Thanks for the contribution @aryannzzz, you'll be credited.
